pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/MethodReturnsInternalArray.xml

```diff
@@ -366,6 +366,52 @@ public class OuterClass {
             return Arrays.copyOf(titles, titles.length);
         }
     }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Nested methods and local class</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class OuterClass {
+    public int[] arrayReturningMethod() {
+        class LocalClass {
+            private String s;
+            public String getString() {
+                return this.s;
+            }
+        }
+        LocalClass c = new LocalClass();
+        return new int[0];
+    }
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Detect returned static arrays</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>16</expected-linenumbers>
+        <code><![CDATA[
+import java.util.concurrent.Callable;
+
+public class MyClass {
+    private static final String[] FOO_BAR = new String[] { "foo", "bar" };
+    private final Callable<String[]> returnsFooBar = new Callable<String[]>() {
+        @Override
+        public String[] call() {
+            return FOO_BAR; // no violation, because FOO_BAR is final. See https://sourceforge.net/p/pmd/bugs/1475/
```

Comment:
In fact this is only safe is the array has length zero... right? We could check the initializer for `{}` and `new X[0]`.

Reply:
That's correct, I guess. I'm hesitating a bit, fixing this bug with this PR. Maybe we should create an issue for this an fix in separately? I think, the bug applies also for pmd6 and fix for https://sourceforge.net/p/pmd/bugs/1475/ needs to be reworked...

Reply:
I think you're right that this needs a ticket to track the issue on master, but it's probably simpler to fix it in this PR for pmd 7 (as a fix on master would need to be adapted to te

Reply:
-> #3160